core/primitives/src/stateless_validation.rs

```diff
@@ -362,3 +370,30 @@ impl ChunkValidatorAssignments {
         }
     }
 }
+
+fn decompress_with_limit(data: &[u8], limit: usize) -> std::io::Result<Vec<u8>> {
+    let mut buf = Vec::new().limit(limit).writer();
+    zstd::stream::copy_decode(data, &mut buf)?;
```

Comment:
what happens if the raw data size is larger than this buffer's size? does decompression fails (I guess ? at the end of the line is for that) ?

Reply:
Yes, in this case an error will be returned here. I've updated that code to have more explicit error message when the limit is exceeded, you can check the complete error message in the `decompress_exceed_limit` test.